refactor of IP service, add ip_test, add projects_test

t0mk opened the pull request:
hi, I refactored the IP service a bit so that it captures the current state of the API.

I identify 2 resources under the /ips/ method:
1) Reservations, which are created by the POST to `/projects/{id}/ips`. It's actually a CIDR block of addresses. 
2) Assignments, which are created by POST to `/devices/{id}/ips`. Those represent coupling between address from a block, and a device.

I noticed that:
- GET to `projects/{id}/ips` will list only reservations, not assignments (but assigmnents are linked from reservations).
- fields which are only in Reservation: facility, available, assignments, addon, bill
- fields which are only in Assignment: assigned_to.

I tried to express that in the golang types without duplication, and then rewrote the API functions with the new types.

I removed the IP listing function from projects.go, as it's already in ip.go. 

I wrote tests for projects.go, and for ip.go. 

This might break some code from https://godoc.org/github.com/packethost/packngo?importers, but I think it's not big deal - most of that is Terraform (which uses govendor, so should be safe), and the other projects use mostly the Devices service, not the IPs.

Please take a look and let me know what do you think.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/packethost/packngo/32)
<!-- Reviewable:end -->

t0mk commented:
The tests actually need Packet API token, as they create resources in Packet.

If you'd like to run the tests from the Drone, you'd need to export PACKET_AUTH_TOKEN, and PACKNGO_TEST=yeah.

mmlb commented:
This is a quick look/first pass.  I'll take a better look a little later today.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ip.go, line 55 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4AOKjEqehs3s6MiIK:-Ks4AOKjEqehs3s6MiIL:bb6t6ux) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/ip.go#L55)):*
> ```Go
> // AvailableRequest is a type for listing available addresses from a reserved block.
> type AvailableRequest struct {
> 	Cidr int `json:"cidr"`
> ```

CIDR

---

*[packngo.go, line 97 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4AxSbtyzgpzRRO1yc:-Ks4AxSbtyzgpzRRO1yd:b2nhgjr) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo.go#L97)):*
> ```Go
> 	Facilities       FacilityService
> 	OperatingSystems OSService
> 	Ips              IPService
> ```

IPs

---

*[packngo_test.go, line 18 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4BMdqtPVqjvXQIszY:-Ks4BMdqtPVqjvXQIszZ:bouod2y) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo_test.go#L18)):*
> ```Go
> They will likely cost you some credit. If you really want to run
> the tests, please export PACKNGO_TEST env var to nonempty string.`
> 	testProjectPrefix = "TEST_DELME_2d768716_"
> ```

this should be something like `PACKNGO_TESTS_DELME_`

---

*[packngo_test.go, line 34 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4BjxdBTKPdjiN1e_q:-Ks4BjxdBTKPdjiN1e_r:b-5km49x) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo_test.go#L34)):*
> ```Go
> func setupWithProject(t *testing.T) (*Client, string, func()) {
> 	c := setup(t)
> 	p, _, err := c.Projects.Create(&ProjectCreateRequest{Name: testProjectPrefix + randString8()})
> ```

maybe just use a uuid? 
Also, we're going to want to go with spot market for these at least for our CI runs, maybe add an env var `CI_SPOT_MARKET` that defaults true?

---

*[packngo_test.go, line 49 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4DyRaHZTxwT_p_GrT:-Ks4DyRaHZTxwT_p_GrU:b3gtp4r) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo_test.go#L49)):*
> ```Go
> 
> func setup(t *testing.T) *Client {
> 	if os.Getenv("PACKNGO_TEST") == "" {
> ```

I think a better name makes sense here, `PACKNGO_TEST_ACTUAL_API`...? idk, does tf have something they use?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32#-:-Ks4EUdoI3kHdZK6rAvU:b-r9cioz)*
<!-- Sent from Reviewable.io -->

t0mk commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[packngo_test.go, line 34 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4BjxdBTKPdjiN1e_q:-Ks8Ie2_3Bm5fbt2wAiZ:byy3548) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo_test.go#L34)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

maybe just use a uuid? 
Also, we're going to want to go with spot market for these at least for our CI runs, maybe add an env var `CI_SPOT_MARKET` that defaults true?
</blockquote></details>

there's no uuidgen in golang standard lib, I figured it's better to have 10 line function than adding a dependency.

I understood only devices are available through the spot market. The tests in this PR only create projects and ip reservations. I will remember to use the spot market when I write device tests.

---

*[packngo_test.go, line 49 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4DyRaHZTxwT_p_GrT:-Ks8LIfE4RohGEZ9O2kn:bj57wzj) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/packngo_test.go#L49)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I think a better name makes sense here, `PACKNGO_TEST_ACTUAL_API`...? idk, does tf have something they use?
</blockquote></details>

Good point, I will change the env var name. In Terraform you need to set TF_ACC=1, "acc" stands for acceptance tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

t0mk commented:
@mmlb I fixed some of your review comments. In order for the tests to run, 2 env vars must be exported: PACKNGO_TEST_ACTUAL_API=1, and PACKET_AUTH_TOKEN.

mmlb commented:
Review status: 3 of 7 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[ip.go, line 16 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH-H6GOlvhkF2bri6s:-KsH-H6GOlvhkF2bri6t:b-1n19zz) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L16)):*
> ```Go
> 	Unassign(assignmentID string) (*Response, error)
> 	GetReservation(reservationID string) (*IPAddressReservation, *Response, error)
> 	GetReservationByCIDR(projectID, cidrString string) (*IPAddressReservation, *Response, error)
> ```

string in cidrString is redundant

---

*[ip.go, line 22 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH1vWFqDaEM83w_VI5:-KsH1vWFqDaEM83w_VI6:bfo4tjc) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L22)):*
> ```Go
> 	RemoveReservation(ipReservationID string) (*Response, error)
> 	GetAvailableAddresses(ipReservationID string, r *AvailableRequest) ([]string, *Response, error)
> }
> ```

any thoughts on the following names instead?
```
Assign
Unassign
Assignment

Reservation
ReservationByCIDR
Reservations
RequestReservations
RemoveReservations
AvailableAddresses
```

It seems like maybe we should have 2 interfaces, `DeviceIPService` and `ProjectIPService` maybe? I'd rather not have overly fat interfaces. If we do 2 interfaces then we can probably also rename
```
RequestReservations -> Request
RemoveReservations -> Remove
```

---

*[ip.go, line 38 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH3nZSiDsN4ieNMBAQ:-KsH3nZTV6di8WzLnW_u:bsdx9hb) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L38)):*
> ```Go
> }
> 
> // IPAddressReservation is created when user sends IP reservation request for his/her project (considering it's within quota).
> ```

his/her -> a

---

*[ip.go, line 48 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH3gdLZdaYDxJMxJ8D:-KsH3gdLZdaYDxJMxJ8E:byooyxl) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L48)):*
> ```Go
> }
> 
> // AvailableResponse is a type for listing of avaialable addresses from a reserved block.
> ```

typo avaia -> avai

---

*[ip.go, line 64 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHBjh9Twfm-7pmTwaa:-KsHBjh9Twfm-7pmTwab:b5vymkm) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L64)):*
> ```Go
> }
> 
> // AddressField is a type for request/response with dict of type {"address": ... }
> ```

I think `type Address struct` seems better, its a little repetitive but not worse than `AddressField.Address`, and `AddressField` indicates to me that its going to be embedded into a struct not stand alone on its own.
Also, comment could use rewording maybe `Address is a type to denote a bare IP address`.
We could also just go with something like:

```
type  Address string

func (a Address) MarshalJSON() ([]byte, error) {
    return fmt.Sprintf(`{"address":"%s"}`, a)
}
```
and we avoid the stuttering

---

*[ip.go, line 78 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHID01bDjk-v5mxoHM:-KsHID01bDjk-v5mxoHN:b-nl8we9) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L78)):*
> ```Go
> 
> // IPServiceOp implements IPService
> type IPServiceOp struct {
> ```

I wonder if it's too late to make this private? We could have a method on `Client` like `Client.IPService()` and it just returns ipServiceOP as a `IPService` (substitute DeviceIPService, ProjectIPService if we go that route.

---

*[ip.go, line 118 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHI-JZA1QVOpbBAEae:-KsHI-JZA1QVOpbBAEaf:b-720r12) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L118)):*
> ```Go
> }
> 
> type ipReservationRoot struct {
> ```

this should be brought into the `ListReservations` func

---

*[ip.go, line 144 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH61YD50yQ097gbRdf:-KsH61YD50yQ097gbRdg:b2eyl9w) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L144)):*
> ```Go
> 	cidrSlice := strings.Split(cidrString, "/")
> 	if len(cidrSlice) != 2 {
> 		return nil, nil, fmt.Errorf("Invalid CIDR expression: %s", cidrString)
> ```

should be `invalid`

---

*[ip.go, line 157 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH7mqmPnrxvuPnrg_K:-KsH7mqmPnrxvuPnrg_L:bozxssi) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L157)):*
> ```Go
> 	for _, r := range rs {
> 		if r.Network == network && r.CIDR == cidr {
> 			return &r, resp, nil
> ```

this is going to allocate r on the heap and and return a pointer to it when I think we just want `&rs[$n]`
can be done like
```go
for i, r := range rs {
...
        return &rs[i], resp, nil
...
```

---

*[ip.go, line 160 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH63ddTxvS6nehcBO0:-KsH63ddTxvS6nehcBO1:b-u8vdth) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L160)):*
> ```Go
> 		}
> 	}
> 	return nil, resp, fmt.Errorf("Couldn't find reservation for CIDR %s", cidrString)
> ```

`couldn't`

---

*[ip.go, line 176 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8ZaObOlERTxQIFPQ:-KsH8ZaObOlERTxQIFPR:b-hekdm5) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L176)):*
> ```Go
> }
> 
> // Unassign unassigns an IP address from the device to which it is currently assignmed.
> ```

typo at the end there, assignmed has an `m` that snuck in

---

*[ip.go, line 183 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8lcGfUSJOAlyj6PK:-KsH8lcGfUSJOAlyj6PL:b-yu2iht) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L183)):*
> ```Go
> }
> 
> // Assign assigns an IP address to a device. The IP address must be in one of the IP ranges assigned to the device’s project.
> ```

break into 2 lines at the `.`

---

*[ip.go, line 206 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8x-UrirmaFphs8gW:-KsH8x-UrirmaFphs8gX:bagpbbg) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L206)):*
> ```Go
> }
> 
> // RequestReservation requests more IP space for a project in order to have additional IP addresses to assign to devices
> ```

missing `.` at the end

---

*[ip.go, line 210 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH94Kb-3_OKktY2Afv:-KsH94Kb-3_OKktY2Afw:bmvf94s) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L210)):*
> ```Go
> 	path := fmt.Sprintf("%s/%s%s", projectBasePath, projectID, ipBasePath)
> 
> 	req, err := i.client.NewRequest("POST", path, &ipReservationReq)
> ```

arg to `client.NewRequest` is pointer to pointer to `IPReservervationRequest`. I'd prefer if ipReservationReq were not even a pointer since we are not going to modify it.

---

*[ip.go, line 229 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHBFROmItMuKhm0zVA:-KsHBFROmItMuKhm0zVB:b-2d9iuc) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L229)):*
> ```Go
> 
> // GetAvailableAddresses lists addresses available from a reserved block
> func (i *IPServiceOp) GetAvailableAddresses(ipReservationID string, r *AvailableRequest) ([]string, *Response, error) {
> ```

ditto pointer -> plain struct

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

t0mk commented:
Review status: 3 of 7 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[ip.go, line 55 at r1](https://reviewable.io:443/reviews/packethost/packngo/32#-Ks4AOKjEqehs3s6MiIK:-KsJ5mxq7MKSdbUfhPcS:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/ba2fdceffb683af940d3e3f40a4a38bf1aad17bc/ip.go#L55)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

CIDR
</blockquote></details>

Done.

---

*[ip.go, line 16 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH-H6GOlvhkF2bri6s:-KsJ5OkSL9hUm9MdihLT:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L16)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

string in cidrString is redundant
</blockquote></details>

Done.

---

*[ip.go, line 22 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH1vWFqDaEM83w_VI5:-KsJ5RahvLj3U_EkOXIM:b-7z08wj) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L22)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

any thoughts on the following names instead?
```
Assign
Unassign
Assignment

Reservation
ReservationByCIDR
Reservations
RequestReservations
RemoveReservations
AvailableAddresses
```

It seems like maybe we should have 2 interfaces, `DeviceIPService` and `ProjectIPService` maybe? I'd rather not have overly fat interfaces. If we do 2 interfaces then we can probably also rename
```
RequestReservations -> Request
RemoveReservations -> Remove
```
</blockquote></details>

I split the methods to the two interfaces as you suggested.

---

*[ip.go, line 38 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH3nZSiDsN4ieNMBAQ:-KsJ5lTHLCdF_05axpGZ:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L38)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

his/her -> a
</blockquote></details>

Done.

---

*[ip.go, line 48 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH3gdLZdaYDxJMxJ8D:-KsJ5mZPV6aQSiOYoIDA:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L48)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

typo avaia -> avai
</blockquote></details>

Done.

---

*[ip.go, line 64 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHBjh9Twfm-7pmTwaa:-KsJ5pfDyDI7UQlHY5s2:b-bxwes8) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L64)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I think `type Address struct` seems better, its a little repetitive but not worse than `AddressField.Address`, and `AddressField` indicates to me that its going to be embedded into a struct not stand alone on its own.
Also, comment could use rewording maybe `Address is a type to denote a bare IP address`.
We could also just go with something like:

```
type  Address string

func (a Address) MarshalJSON() ([]byte, error) {
    return fmt.Sprintf(`{"address":"%s"}`, a)
}
```
and we avoid the stuttering
</blockquote></details>

I changed the name to AddressStruct (Address is taken in facilities.go). I think it's better to keep it as a simple struct rather than define {un}marshaling method: this way it's shorter, cleaner, and consistent with parsing HTTP in other places in the module.

As for the comment: It actually _is_ nothing more than a helper struct to parse simple payload in/to HTTP request/response. 

---

*[ip.go, line 78 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHID01bDjk-v5mxoHM:-KsJ6lXx1_DuWSQHLTh9:b64q9jc) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L78)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I wonder if it's too late to make this private? We could have a method on `Client` like `Client.IPService()` and it just returns ipServiceOP as a `IPService` (substitute DeviceIPService, ProjectIPService if we go that route.
</blockquote></details>

I don't have any opinion on this, but as you wrote, it would have to be changed in all the services, so I think it's beyond the scope of this PR.

---

*[ip.go, line 118 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHI-JZA1QVOpbBAEae:-KsJ7I5HrLgNj73tBkKu:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L118)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

this should be brought into the `ListReservations` func
</blockquote></details>

Done.

---

*[ip.go, line 144 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH61YD50yQ097gbRdf:-KsJ7IueNz2B_s1sEeCB:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L144)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

should be `invalid`
</blockquote></details>

Done.

---

*[ip.go, line 157 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH7mqmPnrxvuPnrg_K:-KsJ7KUCK0sHhdjSgTue:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L157)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

this is going to allocate r on the heap and and return a pointer to it when I think we just want `&rs[$n]`
can be done like
```go
for i, r := range rs {
...
        return &rs[i], resp, nil
...
```
</blockquote></details>

Done.

---

*[ip.go, line 160 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH63ddTxvS6nehcBO0:-KsJ7Kwv0RIAPVjAZjMA:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L160)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

`couldn't`
</blockquote></details>

Done.

---

*[ip.go, line 176 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8ZaObOlERTxQIFPQ:-KsJ7Le8KSgdR9RlZkgr:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L176)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

typo at the end there, assignmed has an `m` that snuck in
</blockquote></details>

Done.

---

*[ip.go, line 183 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8lcGfUSJOAlyj6PK:-KsJ7M2puQ77aDWnsH80:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L183)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

break into 2 lines at the `.`
</blockquote></details>

Done.

---

*[ip.go, line 206 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH8x-UrirmaFphs8gW:-KsJ7Mfu1tyEoZ1f0XkT:b-896fix) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L206)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

missing `.` at the end
</blockquote></details>

Done.

---

*[ip.go, line 210 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH94Kb-3_OKktY2Afv:-KsJ7Ohy4PGv_E9vJsvj:b-atshpc) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L210)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

arg to `client.NewRequest` is pointer to pointer to `IPReservervationRequest`. I'd prefer if ipReservationReq were not even a pointer since we are not going to modify it.
</blockquote></details>

I agree, but at the moment all the structs for create requests for all services are passed by pointer. I can't change it only in the IP services. It's beyond the scope of this PR.

---

*[ip.go, line 229 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHBFROmItMuKhm0zVA:-KsJ7WPdCjLD6Quha-XQ:b3jn4xk) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L229)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

ditto pointer -> plain struct
</blockquote></details>

dtto my comment above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: 3 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ip.go, line 64 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHBjh9Twfm-7pmTwaa:-KsJC6hHOJsUbj7Kxjrs:b-e8rxfk) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L64)):*
<details><summary><i>Previously, t0mk ( Tomáš Karásek) wrote…</i></summary><blockquote>

I changed the name to AddressStruct (Address is taken in facilities.go). I think it's better to keep it as a simple struct rather than define {un}marshaling method: this way it's shorter, cleaner, and consistent with parsing HTTP in other places in the module.

As for the comment: It actually _is_ nothing more than a helper struct to parse simple payload in/to HTTP request/response. 
</blockquote></details>

Right if it is *nothing more* than just json helper then why create a new type for it in the first place? Especially when there's no validation going on at all either. I'd just as well take/return a string and let clients just use that. Struggling with the type name is soft indicator of smell here, AddressField is bad because you'd expect that to be inside a struct somewhere, struct in AddressStruct is redundant...

But I can live with this. I'll keep an eye out to see about changing it across the board later on (... or not).

---

*[ip.go, line 78 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsHID01bDjk-v5mxoHM:-KsJBlk3NVhjSWvmFudH:b-sqdiah) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L78)):*
> Unassign
Yeah as it stands right now we are commenting on the implementation with information that is really about the Interface, kinda smells. But I agree it's definitely out of scope for this PR.

---

*[ip.go, line 210 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH94Kb-3_OKktY2Afv:-KsJ9BN-fj61OqKeTW2x:b-wv7cey) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L210)):*
> AddressStruct
Thats fair, but you don't need to add the `&` when calling into `client.NewRequest` either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 4 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

t0mk commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[ip.go, line 210 at r2](https://reviewable.io:443/reviews/packethost/packngo/32#-KsH94Kb-3_OKktY2Afv:-KsJFfFHg14tBk6SJaB4:bdyi5sc) ([raw file](https://github.com/packethost/packngo/blob/6eb4fc95d6c58e0f473cb185407fe709a812ec05/ip.go#L210)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

> AddressStruct
Thats fair, but you don't need to add the `&` when calling into `client.NewRequest` either.
</blockquote></details>

well spotted! Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 1 of 1 files at r4, 4 of 4 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[ip.go, line 118 at r6](https://reviewable.io:443/reviews/packethost/packngo/32#-KsJYqhG-c3DA2ib6BXG:-KsJYqhG-c3DA2ib6BXH:b-5as279) ([raw file](https://github.com/packethost/packngo/blob/141e7ddebb058ce91048714558e4c458c7d73a49/ip.go#L118)):*
> ```Go
> 	path := fmt.Sprintf("%s/%s%s", deviceBasePath, deviceID, ipBasePath)
> 
> 	req, err := i.client.NewRequest("POST", path, assignRequest)
> ```

err is being ignored

---

*[ip_test.go, line 65 at r6](https://reviewable.io:443/reviews/packethost/packngo/32#-KsJ_78E84Dj3OPTqSLC:-KsJ_78E84Dj3OPTqSLD:blx7cu4) ([raw file](https://github.com/packethost/packngo/blob/141e7ddebb058ce91048714558e4c458c7d73a49/ip_test.go#L65)):*
> ```Go
> 	}
> 
> 	availableAddresses, _, err := c.ProjectIPs.AvailableAddresses(
> ```

err is ignored

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects_test.go, line 15 at r6](https://reviewable.io:443/reviews/packethost/packngo/32#-KsJa_y46H1tb8dEaRnz:-KsJa_y46H1tb8dEaRo-:b-r8je0p) ([raw file](https://github.com/packethost/packngo/blob/141e7ddebb058ce91048714558e4c458c7d73a49/projects_test.go#L15)):*
> ```Go
> 	p, _, err := c.Projects.Create(&pcr)
> 	if err != nil {
> 		t.Error(err)
> ```

t.Error/t.Errorf should probably be Fatal/Fatalf otherwise test execution will continue and we may panic. Probably all instances in both files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 1 of 2 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/32)*
<!-- Sent from Reviewable.io -->
